docs/internal/DistributedArchitectureGuide.md

```diff
@@ -12,6 +12,214 @@ A guide to the general Elasticsearch components can be found [here](https://gith
+
+Incremental bulk indexing has a back-pressure feature. When memory pressure
+grows high we stop reading bytes from TCP sockets for some connections and allow
+only few to proceed, until pressure is resolved. This feature protects from
+unbounded memory usage and OOMs.
+
+ES supports multiple Content-Types for payload: CBOR, JSON, SMILE, YAML, and
+their versioned types. Internally we call it XContentType. When a class
+implements `toXContent` that means it can be sent(serialized) over HTTP/REST.
+
+HTTP routing is based on a combination of Method and URI. For example,
+`RestCreateIndexAction` handler uses `("PUT", "/{index}")`, where curly braces
+indicate path variables. `RestBulkAction` specifies a list of routes
+
+```java
+    @Override
+    public List<Route> routes() {
+        return List.of(
+            new Route(POST, "/_bulk"),
+            new Route(PUT, "/_bulk"),
+            new Route(POST, "/{index}/_bulk"),
+            new Route(PUT, "/{index}/_bulk")
+        );
+    }
+```
+
+Every REST handler must be declared in `ActionModule` class in the
+`initRestHandlers` method. Plugins that implements `ActionPlugin` can extend
+list of handlers through `getRestHandlers` override. Every REST handler should
+extend `BaseRestHandler`.
+
+The job of the REST handler is to parse and validate HTTP request and construct
+typed version of request, often Transport request (see Transport section below).
+
+Request handling flow from Java classes view goes as:
+
+```
+(if security enabled) Security.getHttpServerTransportWithHeadersValidator
+  -> `Netty4HttpServerTransport`
+  -> `AbstractHttpServerTransport`
+  -> `RestController`
+  -> `BaseRestHandler`
+  -> `Rest{Some}Action`
+```
+
+Where `Netty4HttpServerTransport` is a single implementation of
+`AbstractHttpServerTransport` that lives in a `transport-netty4` module. And
+security module injects SSL and headers validator.
+
+## Transport
+
+`Transport` is an umbrella term for a node-to-node communication. It's a
+TCP-based custom binary protocol. Every node in a cluster is a client and server
+at the same time. Node-to-node communication never uses HTTP transport.
+
+`Netty4Transport` is the only implementation of TCP transport and initialize
+Transport client and server. X-pack/security plugin provides secure version of
+transport with TLS and authentication - `SecurityNetty4Transport`.
+
+Once node discovers cluster it will open a pool of connections to every other
+node in a cluster, and every other node will open a pool of connections
+back. That means a connection between nodes A and B is A->B pool + B->A pool. A
+node sends requests only on connections it opens (as a client).
+
+A default connection pool is around 13 connections, a pool has sub-pools of
+connections for different purposes: ping, node-state, bulks, etc. Pool structure
+is defined in `ConnectionProfile` class.
+
+ES has resilience to disconnects, frequent reconnects, but in general we assume
```

Comment:
> resilience

I don't think we are particularly resilient to frequent reconnects. At least, not without being more specific. ES never behaves _incorrectly_ (e.g. loses data) in the face of network outages but it may become unavailable unless the network is stable.